plugins/updater/src/updater.rs

```diff
@@ -540,12 +540,71 @@ impl Update {
         // shouldn't drop but we should be able to pass the reference so we can drop it once the installation
         // is done, otherwise we have a huge memory leak.
 
-        let tmp_dir = tempfile::Builder::new().tempdir()?.into_path();
-        let archive = Cursor::new(bytes);
-        let mut extractor = zip::ZipArchive::new(archive)?;
-        extractor.extract(&tmp_dir)?;
+        // we support 2 type of files exe & msi for now
+        // If it's an `exe` we expect an NSIS installer.
+        enum UpdaterType {
+            Nsis {
+                path: PathBuf,
+                // For extending temp file's life time
+                #[allow(unused)]
+                temp_path: Option<tempfile::TempPath>,
+            },
+            Msi {
+                path: PathBuf,
+                // For extending temp file's life time
+                #[allow(unused)]
+                temp_path: Option<tempfile::TempPath>,
+            },
+        }
 
-        let paths = fs::read_dir(&tmp_dir)?;
+        let updater = 'updater: {
+            #[cfg(feature = "zip")]
+            {
+                if is_zip(&bytes) {
```

Comment:
This can be collapsed
```suggestion
            #[cfg(feature = "zip")]
            if is_zip(&bytes) {
```